---
title: "Lab 7: RNA-Seq workflow: gene-level exploratory analysis and differential expression"
author: "Julia Sousa"
date: "10/26/2020"
output: html_document
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

library(tidyverse)
library(tidyr)

```


```{r, eval=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.11")

```


```{r, eval=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("rnaseqGene")

```


```{r, eval=FALSE}

BiocManager::install("airway")
BiocManager::install("tximeta")
BiocManager::install("Gviz")
BiocManager::install("sva")
BiocManager::install("RUVSeq")
BiocManager::install("fission")

```

```{r}

library(DESeq2)

```


```{r}

library(tidyverse)
library(ggplot2)
library(dplyr)

```


# 2.3 Reading in data with tximeta


```{r}

library("airway")

```

```{r}
dir <- system.file("extdata", package="airway", mustWork=TRUE)

```


```{r}

list.files(dir)

```

```{r}

list.files(file.path(dir, "quants"))

```

```{r}

csvfile <- file.path(dir, "sample_table.csv")
coldata <- read.csv(csvfile, row.names=1, stringsAsFactors=FALSE)
coldata

```

```{r}

coldata <- coldata[1:2,]
coldata$names <- coldata$Run
coldata$files <- file.path(dir, "quants", coldata$names, "quant.sf.gz")
file.exists(coldata$files)

```

```{r}

library("tximeta")

se <- tximeta(coldata)

```

Note: when trying to run this line of code, it asked which directory BiocFileCache directory should access and save tTxDb objects. I chose to send it to what I believe is my local computer. After choosing this option, it did not produce the same output as the tutorial.

```{r}

dim(se)

```

```{r}

head(rownames(se))

```

```{r}

gse <- summarizeToGene(se)

```

```{r}

dim(gse)

```

```{r}

head(rownames(gse))

```


# 2.4 DESeq2 import functions

Notes:
- use different packages to produce desired outputs, along with the function that you need to use
- tximport function, output: list of matrices
- tximeta function, output: SummarizedExperiment
- htseq-count function, output: files
- featureCounts function, output: matrix
- summarizeOverlaps function, output:  SummarizedExperiment

# 2.5 SummarizedExperiment

Notes:
- assay contains matrix of counts
- rowRanges contains info about genomic ranges
- colData contains info about the samples, lines up with the first column of the assay

- tximeta creates an object, gse with three matrices

```{r}

data(gse)
gse

```

```{r}

assayNames(gse)

```

```{r}

head(assay(gse), 3)

```

```{r}

colSums(assay(gse))

```

```{r}

rowRanges(gse)

```

```{r}

seqinfo(rowRanges(gse))

```

```{r}

colData(gse)

```

# 2.6 Branching point

Notes:
- only need SummarizedExperiment object to start analysis


#   The DESeqDataSet object, sample information and the design formula

Notes:
- SummarizedExperiment can be used to move data between packages, automatically subset or reorder the rows or columns associated with rowRanges and colData

```{r}

gse$donor

```

```{r}

gse$condition

```

```{r}

gse$cell <- gse$donor
gse$dex <- gse$condition

```

```{r}

levels(gse$dex)

```
(preserving order:)

```{r}

levels(gse$dex) <- c("untrt", "trt")

```

```{r}

library("magrittr")
gse$dex %<>% relevel("untrt")
gse$dex

```

Note:
- %<>% is the pip for magrittr package

```{r}

gse$gex <- relevel(gse$dex, "untrt")

```


# 3.1 Starting from SummarizedExperiment

```{r}

round( colSums(assay(gse)) / 1e6, 1 )

```

```{r}

library("DESeq2")
dds <- DESeqDataSet(gse, design = ~ cell + dex)

```


# 3.2 Starting from count matrices

```{r}

countdata <- round(assays(gse)[["counts"]])
head(countdata, 3)

```

Note:
- each row represents a gene
- each column represents a sequences RNA library
- vaues give the estimated counts of fragments that were probabilistically assigned to the respective gene in each library
- !! important to check that the columns of the count matrix correspond to the rows of the sample info table manually !!

```{r}

coldata <- colData(gse)

```

Note:
- countdata: table with fragment counts
- coldata: table with info about the samples

```{r}

ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData = coldata,
                                 design = ~ cell + dex)

```

# 4 Exploratory analysis and visualization


# 4.1 Pre-filtering the dataset

```{r}

nrow(dds)

```

```{r}

keep <- rowSums(counts(dds)) > 1
dds <- dds[keep,]
nrow(dds)

```

Note:
- some datasets need more filtering
  - can specify that at least 3 samples have a count of 10 or higher

```{r}

keep <- rowSums(counts(dds) >= 10) >= 3

```


# 4.2 The variance stabilizing transformation and the rlog

```{r}

lambda <- 10^seq(from = -1, to = 2, length = 1000)
cts <- matrix(rpois(1000*100, lambda), ncol = 100)
library("vsn")
meanSdPlot(cts, ranks = FALSE)

```

```{r}

log.cts.one <- log2(cts + 1)
meanSdPlot(log.cts.one, ranks = FALSE)

```

```{r}

vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)

```

```{r}

colData(vsd)

```

```{r}

rld <- rlog(dds, blind = FALSE)
head(assay(rld), 3)

```

Note: 
- "blind = FALSE" means that differences between cell lines and treatment won't contribute to the expected variance-mean trend of the experiment

```{r}

library("dplyr")
library("ggplot2")

dds <- estimateSizeFactors(dds)

df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized = TRUE)[, 1:2] + 1)) %>% 
    mutate(transformation = "log2(x + 1"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"))

colnames(df)[1:2] <- c("x", "y")

lvls <- c("log2(x + 1)", "vst", "rlog")
df$transformation <- factor(df$transformation, levels = lvls)

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid(. ~ transformation)

```


# 4.3 Sample distances

```{r}

sampleDists <- dist(t(assay(vsd)))
sampleDists

```

```{r}

library("pheatmap")
library("RColorBrewer")

```

Note:
- must manually provide sampleDists to culstering_distance of pheatmap other pheatmap will assume matrix is the data values themselves and caluculate distances between rows/columns of distance matrix (not what is wanted)

```{r}

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$dex, vsd$cell, sep = " -  ")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) ) (255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)

```

- can also calculate sample distances using Poisson Distance (PoiClaClu package)

```{r}

library("PoiClaClu")
poisd <- PoissonDistance(t(counts(dds)))

```

```{r}

samplePoisDistMatrix <- as.matrix( poisd$dd )
rownames(samplePoisDistMatrix) <- paste( dds$dex, dds$cell, sep=" - " )
colnames(samplePoisDistMatrix) <- NULL
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = colors)

```



# 4.4 PCA plot

```{r}

plotPCA(vsd, intgroup = c("dex", "cell"))

```

```{r}

pcaData <- plotPCA(vsd, intgroup = c("dex", "cell"), returnData = TRUE)
pcaData

```

```{r}

percentVar <- round(100 * attr(pcaData, "percentVar"))

```

```{r}

ggplot(pcaData, aes(x = PC1, y = PC2, color = dex, shape = cell)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA with VST data")

```


# 4.5 PCA plot using Generalized PCA


Note: 
- can use generalized principal component analysis (GLM-PCA) to perform dimension reduction on data that's not normally distributed


```{r}

library("glmpca")

gpca <- glmpca(counts(dds), L = 2)
gpca.dat <- gpca$factors
gpca.dat$dex <- dds$dex
gpca.dat$cell <- dds$cell

ggplot(gpca.dat, aes(x = dim1, y = dim2, color = dex, shape = cell)) + 
  geom_point(size = 3) + coord_fixed() + ggtitle("glmpca - Generalized PCA")

```


# 4.6 MDS plot

Note: 
- multidimensional scaling (MDS) is useful when don't have a matrix of data, but only a matrix of distances

```{r}

mds <- as.data.frame(colData(vsd))  %>%
         cbind(cmdscale(sampleDistMatrix))
ggplot(mds, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with VST data")

```

```{r}

mdsPois <- as.data.frame(colData(dds)) %>%
   cbind(cmdscale(samplePoisDistMatrix))
ggplot(mdsPois, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with PoissonDistances")

```

# 5 Differentiala expression analysis

# 5.1 Running differential expression pipeline

```{r}

dds <- DESeq(dds)

```


# 5.2 Building the results table

```{r}

res <- results(dds)
res

```

```{r}

res <- results(dds, contrast = c("dex", "trt", "untrt"))

```


```{r}

mcols(res, use.names = TRUE)

```

```{r}

summary(res)

```

```{r}

res.05 <- results(dds, alpha = 0.05)
table(res.05$padj < 0.05)

```

```{r}

resLFC1 <- results(dds, lfcThreshold = 1)
table(resLFC1$padj < 0.1)

```


# 5.3 Other comparisions

```{r}

results(dds, contrast = c("cell", "N061011", "N61311"))

```


# 5.4 Multiple teasing

```{r}

sum(res$pvalue < 0.05, na.rm = TRUE)

```

```{r}

sum(!is.na(res$pvalue))

```

```{r}

sum(res$padj < 0.1, na.rm = TRUE)

```

```{r}

resSig <- subset(res, padj < 0.1)
head(resSig[ order(resSig$log2FoldChange), ])

```

```{r}

head(resSig[ order(resSig$log2FoldChange, decreasing = TRUE), ])

```


# Plotting results

# 6.1 Counts plot

```{r}

topGene <- rownames(res)[which.min(res$padj)]
plotCounts(dds, gene = topGene, intgroup = c("dex"))

```

```{r}

library("ggbeeswarm")

geneCounts <- plotCounts(dds, gene = topGene, intgroup = c("dex", "cell"),
                        returnData = TRUE)
ggplot(geneCounts, aes(x = dex, y = count, color = cell)) +
  scale_y_log10() + geom_beeswarm(cex = 3)

```

```{r}

ggplot(geneCounts, aes(x = dex, y = count, color = cell, group = cell)) +
  scale_y_log10() + geom_point(size = 3) + geom_line()

```


# 6.2 MA-plot

Note:
- MA-plot gives useful overview for distribution of estimated coefficients in the model
- y axis: "M" means "minus" (subtraction of log values is equal to log of the ratio)
- x axis: "A" means "average"

```{r}

library("apeglm")
resultsNames(dds)

```

```{r}

res <- lfcShrink(dds, coef = "dex_trt_vs_untrt", type = "apeglm")
plotMA(res, ylim = c(-5, 5))

```

```{r}

res.noshr <- results(dds, name = "dex_trt_vs_untrt")
plotMA(res.noshr, ylim = c(-5, 5))

```

```{r}

plotMA(res, ylim = c(-5,5))
topGene <- rownames(res)[which.min(res$padj)]
with(res[topGene, ], {
  points(baseMean, log2FoldChange, col="dodgerblue", cex=2, lwd=2)
  text(baseMean, log2FoldChange, topGene, pos=2, col="dodgerblue")
})

```

```{r}

hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white")

```


# 6.3 Gene clustering

```{r}

library("genefilter")
topVarGenes <- head(order(rowVars(assay(vsd)), decreasing = TRUE), 20)

```

```{r, fig.height = 5, fig.width = 5}

mat  <- assay(vsd)[ topVarGenes, ]
mat  <- mat - rowMeans(mat)
anno <- as.data.frame(colData(vsd)[, c("cell","dex")])
pheatmap(mat, annotation_col = anno)

```


# 6.4 Independent filtering

```{r}

qs <- c(0, quantile(resLFC1$baseMean[resLFC1$baseMean > 0], 0:6/6))
bins <- cut(resLFC1$baseMean, qs)
levels(bins) <- paste0("~", round(signif((qs[-1] + qs[-length(qs)])/2, 2)))
fractionSig <- tapply(resLFC1$pvalue, bins, function(p)
                          mean(p < .05, na.rm = TRUE))
barplot(fractionSig, xlab = "mean normalized count",
                     ylab = "fraction of small p values")

```


# 6.5 Independent Hypothesis Weighting

```{r, eval=FALSE}

BiocManager::install("IHW")

```

```{r}

library("IHW")
res.ihw <- results(dds, filterFun = ihw)

```


# 7 Annotating and exporting results

```{r}

library("AnnotationDbi")
library("org.Hs.eg.db")

```

```{r}

columns(org.Hs.eg.db)

```

```{r}

ens.str <- substr(rownames(res), 1, 15)
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

```

```{r}

resOrdered <- res[order(res$pvalue),]
head(resOrdered)

```


# 7.1 Exporting results


```{r, eval = FALSE}

BiocManager::install("ReportingTools")

```

```{r}

resOrderedDF <- as.data.frame(resOrdered)[1:100, ]
write.csv(resOrderedDF, file = "results.csv")

```

Note:
- ReoprtingTools automatically generate dynamic HTML documents

```{r, eval=FALSE}

library("ReportingTools")


htmlRep <- HTMLReport(shortName="report", title="My report",
                 reportDirectory="./report")
publish(resOrderedDF, htmlRep)
url <- finish(htmlRep)
browseURL(url)

```




# 7.2 Plotting fold changes in genomic space

```{r}

resGR <- lfcShrink(dds, coef = "dex_trt_vs_untrt", type = "apeglm", format = "GRanges")
resGR

```

```{r}

ens.str <- substr(names(resGR), 1, 15)
resGR$symbol <- mapIds(org.Hs.eg.db, ens.str, "SYMBOL", "ENSEMBL")

```


```{r, eval=FALSE}
BiocManager::install("Gviz")
```


```{r}

library("Gviz")

```

```{r}

window <- resGR[topGene] + 1e6
strand(window) <- "*"
resGRsub <- resGR[resGR %over% window]
naOrDup <- is.na(resGRsub$symbol) | duplicated(resGRsub$symbol)
resGRsub$group <- ifelse(naOrDup, names(resGRsub), resGRsub$symbol)

```

```{r}

status <- factor(ifelse(resGRsub$padj < 0.05 & !is.na(resGRsub$padj),
                        "sig", "notsig"))

```

```{r}

options(ucscChromosomeNames = FALSE)
g <- GenomeAxisTrack()
a <- AnnotationTrack(resGRsub, name = "gene ranges", feature = status)
d <- DataTrack(resGRsub, data = "log2FoldChange", baseline = 0,
               type = "h", name = "log2 fold change", strand = "+")
plotTracks(list(g, d, a), groupAnnotation = "group",
           notsig = "grey", sig = "hotpink")

```


# 8 Removing hidden batch effects

Notes:
- can detect groupings of the samples using statistical methods designed for RNA-seq (sva package or RUVSeq package)
- SVA package uses "surrogate variables" for the estimated variables that are wanted in analysis
- RUV package uses "factors of unwanted variation" with "Remove Unwanted Variation" (RUV)
- use SVA first to find hidden batch effects, then use RUV



# 8.1 Using SVA with DESeq2

```{r}

library("sva")

```

```{r}

dat <- counts(dds, normalized = TRUE)
idx <- rowMeans(dat) > 1
dat <- dat[idx, ]
mod <- model.matrix(~dex, colData(dds))
mod0 <- model.matrix(~  1, colData(dds))
svseq <- svaseq(dat, mod, mod0, n.sv = 2)

```

```{r}

svseq$sv

```

```{r, fig.height = 3}

par(mfrow = c(2, 1), mar = c(3, 5, 3, 1))
for (i in 1:2) {
  stripchart(svseq$sv[, i] ~ dds$cell, vertical = TRUE, main = paste0("sv", i))
  abline(h = 0)
}

```

```{r}

ddssva <- dds
ddssva$SV1 <- svseq$sv[,1]
ddssva$SV2 <- svseq$sv[,2]
design(ddssva) <- ~ SV1 + SV2 + dex

```



# 8.2 Using RUV with DESeq2

```{r}

library("RUVSeq")

```

```{r}

set <- newSeqExpressionSet(counts(dds))
idx  <- rowSums(counts(set) > 5) >= 2
set  <- set[idx, ]
set <- betweenLaneNormalization(set, which="upper")
not.sig <- rownames(res)[which(res$pvalue > .1)]
empirical <- rownames(set)[ rownames(set) %in% not.sig ]
set <- RUVg(set, empirical, k=2)
pData(set)

```

```{r, fig.height = 3}

par(mfrow = c(2, 1), mar = c(3,5,3,1))
for (i in 1:2) {
  stripchart(pData(set)[, i] ~ dds$cell, vertical = TRUE, main = paste0("W", i))
  abline(h = 0)
 }

```

```{r}

ddsruv <- dds
ddsruv$W1 <- set$W_1
ddsruv$W2 <- set$W_2
design(ddsruv) <- ~ W1 + W2 + dex

```


# 9 Time course experiments

```{r}

library("fission")
data("fission")
ddsTC <- DESeqDataSet(fission, ~ strain + minute + strain:minute)

```

```{r}

ddsTC <- DESeq(ddsTC, test="LRT", reduced = ~ strain + minute)
resTC <- results(ddsTC)
resTC$symbol <- mcols(ddsTC)$symbol
head(resTC[order(resTC$padj),], 4)

```


```{r}

fiss <- plotCounts(ddsTC, which.min(resTC$padj),
                   intgroup = c("minute", "strain"), returnData = TRUE)
fiss$minute <- as.numeric(as.character(fiss$minute))

ggplot(fiss,
       aes(x = minute, y = count, color = strain, group = strain)) + 
  geom_point() + stat_summary(fun.y = mean, geom = "line") +
  scale_y_log10()

```

```{r}

resultsNames(ddsTC)

```

```{r}

res30 <- results(ddsTC, name = "strainmut.minute30", test = "Wald")
res30[which.min(resTC$padj),]

```

```{r}

betas <- coef(ddsTC)
colnames(betas)

```

```{r, fig.height = 5, fig.width = 5}

topGenes <- head(order(resTC$padj), 20)
mat <- betas[topGenes, -c(1,2)]
thr <- 3
mat[mat < -thr] <- -thr
mat[mat > thr] <- thr
pheatmap(mat, breaks = seq(from = -thr, to = thr, length = 101),
         cluster_col = FALSE)

```

# 10 Session information

Note:
- sessionInfo reports the versino of numbers of R and all packages used in this session
- allows to keep record of this and track down what's happened thus far in case R no longer works or gives new results because functions have been changed in a newer package version

```{r}

sessionInfo()

```


















