---
title: "Lab 7: RNA-Seq workflow: gene-level exploratory analysis and differential expression"
author: "Julia Sousa"
date: "10/26/2020"
output: html_document
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

library(tidyverse)

```


```{r}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.11")

```


```{r}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("rnaseqGene")

```


```{r}

BiocManager::install("airway")

```

```{r}

BiocManager::install("tximeta")

```

```{r}
BiocManager::install("DESeq2")
BiocManager::install("Gviz")
BiocManager::install("sva")
BiocManager::install("RUVSeq")
BiocManager::install("fission")

```

```{r}

library(tidyverse)
library(ggplot2)
library(dplyr)

```


# 2.3 Reading in data with tximeta


```{r}

library("airway")

```

```{r}
dir <- system.file("extdata", package="airway", mustWork=TRUE)

```


```{r}

list.files(dir)

```

```{r}

list.files(file.path(dir, "quants"))

```

```{r}

csvfile <- file.path(dir, "sample_table.csv")
coldata <- read.csv(csvfile, row.names=1, stringsAsFactors=FALSE)
coldata

```

```{r}

coldata <- coldata[1:2,]
coldata$names <- coldata$Run
coldata$files <- file.path(dir, "quants", coldata$names, "quant.sf.gz")
file.exists(coldata$files)

```

```{r}

library("tximeta")

se <- tximeta(coldata)

```

Note: when trying to run this line of code, it asked which directory BiocFileCache directory should access and save tTxDb objects. I chose to send it to what I believe is my local computer. After choosing this option, it did not produce the same output as the tutorial.

```{r}

dim(se)

```

```{r}

head(rownames(se))

```

```{r}

gse <- summarizeToGene(se)

```

```{r}

dim(gse)

```

```{r}

head(rownames(gse))

```


# 2.4 DESeq2 import functions

























